tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -414,43 +421,59 @@ void SPIRVEmitter::doFunctionDecl(const FunctionDecl *decl) {
-    const uint32_t ptrType =
-        theBuilder.getPointerType(typeTranslator.translateType(decl->getType()),
-                                  spv::StorageClass::Function);
+  // The contents in externally visible variables can be updated via the
+  // pipeline. They should be handled differently from file and function scope
+  // variables.
+  // File scope variables (static "global" and "local" variables) belongs to
+  // the Private storage class, while function scope variables (normal "local"
+  // variables) belongs to the Function storage class.
+  if (!decl->isExternallyVisible()) {
+    // We already know the variable is not externally visible here. If it does
+    // not has local storage, it should be file scope variable.
+    const bool isFileScopeVar = !decl->hasLocalStorage();
+    const uint32_t varType = typeTranslator.translateType(decl->getType());
 
     // Handle initializer. SPIR-V requires that "initializer must be an <id>
     // from a constant instruction or a global (module scope) OpVariable
     // instruction."
-    llvm::Optional<uint32_t> constInitializer = llvm::None;
-    uint32_t varInitializer = 0;
+    llvm::Optional<uint32_t> constInit;
     if (decl->hasInit()) {
-      const Expr *declInit = decl->getInit();
-
       // First try to evaluate the initializer as a constant expression
       Expr::EvalResult evalResult;
-      if (declInit->EvaluateAsRValue(evalResult, astContext) &&
+      if (decl->getInit()->EvaluateAsRValue(evalResult, astContext) &&
           !evalResult.HasSideEffects) {
-        constInitializer = llvm::Optional<uint32_t>(
+        constInit = llvm::Optional<uint32_t>(
             translateAPValue(evalResult.Val, decl->getType()));
       }
-
-      // If we cannot evaluate the initializer as a constant expression, we'll
-      // need use OpStore to write the initializer to the variable.
-      if (!constInitializer.hasValue()) {
-        varInitializer = doExpr(declInit);
-      }
+    } else if (isFileScopeVar) {
+      // For static variables, if no initializers are provided, we should
+      // initialize them to zero values.
+      constInit = llvm::Optional<uint32_t>(theBuilder.getConstantNull(varType));
     }
 
     const uint32_t varId =
-        theBuilder.addFnVariable(ptrType, decl->getName(), constInitializer);
+        isFileScopeVar
+            ? theBuilder.addFileVar(varType, decl->getName(), constInit)
+            : theBuilder.addFnVar(varType, decl->getName(), constInit);
     declIdMapper.registerDeclResultId(decl, varId);
 
-    if (varInitializer) {
-      theBuilder.createStore(varId, varInitializer);
+    // If we cannot evaluate the initializer as a constant expression, we'll
+    // need use OpStore to write the initializer to the variable.
+    // Also we should only evaluate the initializer once for a static variable.
+    if (decl->hasInit() && !constInit.hasValue()) {
+      if (isFileScopeVar) {
+        if (decl->isStaticLocal()) {
```

Comment:
That's a little bit confusing actually. "With function scope" in the doc is incorrect I think. It's more appropriate to say "within function". Because clearly static local variables are not of function scope.